### PR TITLE
Resume functionality for margin cache pipeline.

### LINKED ISSUE #313: Add resume functionality for margin cache generation
**Feature request**

Similar to catalog import, enable some logging of completed stages, so progress can be resumed for long-running pipelines that are interrupted.

**Before submitting**
Please check the following:

- [x] I have described the purpose of the suggested change, specifying what I need the enhancement to accomplish, i.e. what problem it solves.
- [x] I have included any relevant links, screenshots, environment information, and data relevant to implementing the requested feature, as well as pseudocode for how I want to access the new functionality.
- [ ] If I have ideas for how the new feature could be implemented, I have provided explanations and/or pseudocode and/or task lists for the steps.

### PR DESCRIPTION
**Pull request by delucchi-cmu**
## Change Description

Closes #313 .

Adds a resume plan to margin cache generation.

## Code Quality
- [x] I have read the [Contribution Guide](https://hipscat-import.readthedocs.io/en/stable/guide/contributing.html) and [LINCC Frameworks Code of Conduct](https://lsstdiscoveryalliance.org/programs/lincc-frameworks/code-conduct/)
- [x] My code follows the code style of this project
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

### New Feature Checklist
- [x] I have added or updated the docstrings associated with my feature using the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [x] I have added unit/End-to-End (E2E) test cases to cover my new feature
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### CODE REVIEW

File: src/hipscat_import/margin_cache/margin_cache_resume_plan.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        self.wait_for_futures(futures, self.REDUCING_STAGE)
+        remaining_sources_to_reduce = self.get_remaining_reduce_keys()
+        if len(remaining_sources_to_reduce) > 0:
+            raise RuntimeError(
+                f"{len(remaining_sources_to_reduce)} reducing stages did not complete successfully."
+            )
+        self.touch_stage_done_file(self.REDUCING_STAGE)
+
+
+def _find_partition_margin_pixel_pairs(stats, margin_order):
```
Comment (camposandro) [inline src/hipscat_import/margin_cache/margin_cache_resume_plan.py:141]: Can we change the `stats` argument to `combined_pixels`, and maybe add a sentence explaining what those are in the docstring?

```suggestion
def _find_partition_margin_pixel_pairs(combined_pixels, margin_order):
```

File: src/hipscat_import/margin_cache/margin_cache_resume_plan.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            margin_pix.append(m_p)
+
+    margin_pairs_df = pd.DataFrame(
+        zip(norders, part_pix, margin_pix),
+        columns=["partition_order", "partition_pixel", "margin_pixel"],
+    )
+    return margin_pairs_df
+
+
+def _create_margin_directory(stats, output_path, storage_options):
```
Comment (camposandro) [inline src/hipscat_import/margin_cache/margin_cache_resume_plan.py:169]: Can we apply the same here?

File: src/hipscat_import/margin_cache/margin_cache_map_reduce.py
Comment (camposandro) [inline src/hipscat_import/margin_cache/margin_cache_map_reduce.py:1]: It'd be great to have the failure prints introduced in https://github.com/astronomy-commons/hipscat-import/pull/318 here as well!

File: src/hipscat_import/margin_cache/margin_cache_resume_plan.py (hunk 2 above)
Comment (delucchi-cmu) [reply src/hipscat_import/margin_cache/margin_cache_resume_plan.py]: Nah - I just removed the method. The reduce method creates directories as needed, so we don't need to pre-create them.

File: src/hipscat_import/margin_cache/margin_cache_map_reduce.py
Comment (delucchi-cmu) [reply src/hipscat_import/margin_cache/margin_cache_map_reduce.py:1]: Great catch! Thank you!

**Review by camposandro — APPROVED** [review]

**Pull request merged**